### DISCUSSION TITLE
Merging into monorepos: scaling challenges?

### DISCUSSION THREAD

**Discussion by sprihatucker** [MobileNativeFoundation/discussions#127] [General]
Hey everyone, we’re building [MergeQueue](https://mergequeue.com/): the merge-automation tool for Github that teams use to eliminate build failures on master/main. As we grow, we're looking to learn from other folks’s experience building/ using similar tools. Specifically, thoughts on any of the following:

1. If you’re on a monorepo/ repo with a lot of merges across the team, how do you coordinate your merging process?
2. How do you currently prevent implicit dependencies between PRs from different authors from breaking the main build when they’re merged? If you choose not to, is it because broken master/main doesn’t happen often? Some other reason?
3. Are there any limits to the system you used/are using? Any observations as you scaled the team?
4. At what point did you need to parallelize merge processing (if at all)? 

Any thoughts/ ideas relevant to us are welcome! 

P.S: In case it helps, here’s a [link to our Hacker News launch](https://news.ycombinator.com/item?id=27858013) from a month ago :)

**Comment by keith** [comment]
At Lyft we don't have a company wide monorepo, but our iOS and Android apps do live in their own single repos, so they have quite a few contributors.

We use a rebase-and-merge strategy, this isn't perfect but it generally solves the case of "someone tried to merge an old branch and hadn't ever rebased it". But it misses "2 people merged around the exact same time with conflicting changes", in general we're pretty happy with how this strikes a balance between ease of implementation (literally 1) rebase 2) merge on green) vs being solid enough.

The downside here is that it means 2 CI runs per PR, but the upside is that if a single commit in a train (if we used release trains) failed, there's no complex and time consuming bisecting to determine which commit was bad while blocking the other commits.
